towhee/engine/pipeline.py

```diff
@@ -26,65 +27,91 @@ class Pipeline:
     def __init__(self, graph_repr: GraphRepr, parallelism: int = 1) -> None:
         """
         Args:
-            graph_repr: (yaml `str` or `towhee.dag.GraphRepr`)
-                The graph representation
+            graph_repr: (`str` or `towhee.dag.GraphRepr`)
+                The graph representation either as a YAML-formatted string, or directly
+                as an instance of `GraphRepr`.
             parallelism: (`int`)
                 how many rows of inputs to be processed concurrently
         """
-        if isinstance(graph_repr, str):
-            self._graph_repr = GraphRepr.from_yaml(graph_repr)
-        else:
-            self._graph_repr = graph_repr
 
         self._parallelism = parallelism
         self.on_graph_finish_handlers = []
         self._scheduler = None
 
+        # TODO(fzliu): instantiate `DataFrame` columns.
+        self._outputs = None
+
+        # Instantiate the graph representation.
+        if isinstance(graph_repr, str):
+            self._graph_repr = GraphRepr.from_yaml(graph_repr)
+        else:
+            self._graph_repr = graph_repr
+
         # self.fill_cache()
 
         # self._cv = threading.Condition()
         # self.on_graph_finish_handlers = [self._on_graph_finish]
         # self._build()
 
-    def register(self, scheduler):
+    @property
+    def outputs(self) -> DataFrame:
+        return self._outputs
+
+    def register(self, scheduler: TaskScheduler):
         self._scheduler = scheduler
 
     def __call__(self, inputs: DataFrame) -> DataFrame:
-        """
-        Process one input data
-        """
-        if self._scheduler is None:
-            raise AttributeError(
-                'The pipeline is not registered to a scheduler')
+        """Process an input `DataFrame`.
 
-        assert inputs.size == 1
-        g = GraphContext(0, self._graph_repr)
-        self._scheduler.register(g)
-        _, data = inputs.get(0, 1)
-        g(data[0])
-        return g.result()
+        Args:
+            inputs: (`towhee.DataFrame`)
+        """
+        # Pipelines must register themselves with schedulers.
+        if not self._scheduler:
+            raise AttributeError('Pipeline not registered to a Scheduler.')
+
+        self._outputs = DataFrame()
+
+        # The parallelism parameter dictates how many copies of the graph context we
+        # create. This is likely a low number (1-4) for local engines, but may be much
+        # higher for cloud instances.
+        graph_ctxs = []
+        for n in range(self._parallelism):
+            graph_ctx = GraphContext(n, self._graph_repr)
+            self._scheduler.register(graph_ctx)
+            graph_ctxs.append(graph_ctx)
+
+        # Weave inputs into each GraphContext.
+        # TODO(fzliu): can we spin off a thread to do this?
+        for n, row in enumerate(inputs.map_iter()):
+            idx = n % len(graph_ctxs)
+            graph_ctxs[idx](row[0])
```

Comment:
a graph_ctxs can not process multiple rows simultaneously.
if input rows > len(graph_ctxs), will a graph context being called again before it finish its last call?

Reply:
Made `_start_op` and `_end_op` schedulable operators and added callbacks to set a `GraphContext`'s `_is_busy` flag. When `_is_busy` evaluates to True, new inputs to `_start_op` disabled.